cfgov/scripts/unpublish_closed_jobs.py

```diff
@@ -0,0 +1,74 @@
+import logging
+from urllib2 import urlopen
+import requests
+import json
+import os
+
+from bs4 import BeautifulSoup
+
+from jobmanager.models.pages import JobListingPage
+
+logger = logging.getLogger(__name__)
+
+URL = os.environ.get('USAJOBS_ARCHIVE_URL')
+HEADERS = {
+    'Host': os.environ.get('USAJOBS_HOST'),
+    'User-Agent': os.environ.get('USAJOBS_USER'),
+    'Authorization-Key': os.environ.get('USAJOBS_API_KEY'),
+}
+
+
+def job_page_closed(link):
+    try:
+        webpage = urlopen(link)
+        soup = BeautifulSoup(webpage, 'html.parser')
+        closed_text = soup.find('div', attrs={'class': 'usajobs-joa-closed'})
+        if closed_text:
+            return 'This job announcement has closed' in closed_text.contents
+    except Exception:
+        logger.exception('Check of USAJobs page "{}" failed'.format(link))
```

Comment:
@virginiacc one other thought: should this exception (and the one below) cause the job to fail? Currently if these checks fail there'll be a logging statement, but the command (and, presumably, Jenkins job) will succeed unnoticed.

Reply:
@chosak That's a good question. I think it's possible that one check could fail but others succeed if there are multiple live job postings. Maybe I could restructure to fail the job if none of the checks succeed?

Or could we notify Human Capital of the results of this job? I've been thinking of this as an optimization rather than complete replacement of the current process -- we still need to double-check that jobs have closed on our site in case of API issues -- and reporting the outcome of this script could be the way to do that.

Reply:
@virginiacc if the USAJobs site tends to be down intermittently, and we expect this job to fail from time to time, then it might be confusing/annoying to notify HC every time that happens. If this job is running frequently (hourly?) then it might not matter if any one or several jobs fails.

Following standard Python/command line practice a user might expect that when some error happens the command will exit with a non-zero return code. It might be a good idea to keep checking jobs even if a single one fails, although I'd guess most likely if one fails they all would.

The way I might try to set it up is to have the job fail if any checks fail, and then setup the Jenkins job that calls this to notify the appropriate people after something like two failures in a row. That way if there's some one-off hiccup that isn't a big deal, people aren't bothered with nuisance emails, but if there's a continuous failure, they'll know about it and be able to respond.

Reply:
@chosak The USAJobs API/site seems stable, but I encountered at least one instance in testing where my requests repeatedly timed out for a few minutes, so I wanted to account for that possibility. I updated in https://github.com/cfpb/cfgov-refresh/pull/3787/commits/716a4659ca406a328e589358c9277d773f87ae10 to fail the job if an API or page check results in an exception, and also to target the checks based on applicant type.

Jobs always close at midnight on USAJobs, so I'd been thinking this could run once a night -- but maybe it should re-run at an interval if it fails the first time, and only send notifications if it continues to fail?

Reply:
@virginiacc that sounds good.

Given our current Jenkins conventions I believe it's straightforward to send emails on the second failure. I don't know of a way to alter the interval of running based on success/failure of previous jobs, though. What if you had this job run twice nightly, say, midnight and 1AM? In success cases, the 1AM run would have nothing to do (effectively be a no-op), but in failure cases, it'd be both a re-try and allow for sending of that error email.

Reply:
@chosak That sounds like it would work perfectly!